dmoj/executors/BF.py

```diff
@@ -1,80 +1,215 @@
+            head = f'%p{i}'
+        elif c in '+-':
+            v = 1 if c == '+' else -1
+            inst.append(f'%b{i} = load i8, i8* {head}')
+            inst.append(f'%c{i} = add i8 %b{i}, {v}')
+            inst.append(f'store i8 %c{i}, i8* {head}')
+        elif c == ',':
+            inst.append(f'%b{i} = call i32 @getchar_unlocked()')
+            inst.append(f'%c{i} = trunc i32 %b{i} to i8')
+            inst.append(f'store i8 %c{i}, i8* {head}')
+        elif c == '.':
+            inst.append(f'%b{i} = load i8, i8* {head}')
+            inst.append(f'%c{i} = zext i8 %b{i} to i32')
+            inst.append(f'call i32 @putchar_unlocked(i32 %c{i})')
+        elif c == '[':
+            inst.append(f'br label %y{i}')
+            inst.append(f'x{i}:')
+            inst.append(f'%p{i}x = phi i8* [%p{i}y, %y{i}]')
+            stack.append((i, f'[{head}, {label}]'))
+            head, label = f'%p{i}x', f'%x{i}'
+        elif c == ']':
+            j, phi = stack.pop()
+            inst.append(f'br label %y{j}')
+            inst.append(f'y{j}:')
+            inst.append(f'%p{j}y = phi i8* {phi}, [{head}, {label}]')
+            inst.append(f'%b{j} = load i8, i8* %p{j}y')
+            inst.append(f'%c{j} = icmp ne i8 %b{j}, 0')
+            inst.append(f'br i1 %c{j}, label %x{j}, label %z{j}')
+            inst.append(f'z{j}:')
+            head, label = f'%p{j}y', f'%z{j}'
 
-    def launch(self, *args, **kwargs):
+    assert not stack
+    return TEMPLATE.replace(b'{code}', '\n'.join(inst).encode())
+
+
+OPT_PASSES = [
+    # Shorten instructions
+    'instcombine',
+    'gvn',
+    # Optimize loops
+    'loop-rotate',
+    'loop-mssa(licm)',
+    'indvars',
+    # Clean up
+    'simplifycfg',
+    'instcombine',
+    'gvn',
+]
+
+
+class Executor(LLCExecutor):
+    test_program = ',+[-.,+]'
+
+    def create_files(self, problem_id: str, source_code: bytes, *args, **kwargs) -> None:
+        super().create_files(problem_id, compile_to_llvm(source_code), *args, **kwargs)
+
+    def launch(self, *args, **kwargs) -> TracedPopen:
         memory = kwargs['memory']
         # For some reason, RLIMIT_DATA is being applied to our mmap, so we have to increase the memory limit.
         kwargs['memory'] += 8192
         return super().launch(str(memory * 1024), **kwargs)
 
-    def _has_invalid_brackets(self, source_code) -> bool:
-        open_brackets = 0
-        for c in source_code:
-            if c == b'['[0]:
-                open_brackets += 1
-            elif c == b']'[0]:
-                if open_brackets == 0:
-                    return True
-                else:
-                    open_brackets -= 1
-        return open_brackets != 0
+    # Do both opt and llc.
+    def assemble(self) -> Tuple[bytes, List[str]]:
+        assert self._code is not None
+        # TODO: find an alternative to get opt_path
+        opt_path = self.get_as_path().rstrip(self.as_name) + 'opt'
```

Comment:
I am horrified by this since `rstrip` removes any character that appears in `self.as_name` until it sees one that doesn't. Can you at least do
```suggestion
        opt_path = re.sub(f'{re.escape(self.as_name)}$', 'opt', self.get_as_path())
```

Reply:
Done.